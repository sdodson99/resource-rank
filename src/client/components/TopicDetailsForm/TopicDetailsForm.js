import Link from 'next/link';
import React from 'react';
import ErrorAlert from '../Alerts/ErrorAlert/ErrorAlert';
import LoadingSpinner from '../LoadingSpinner/LoadingSpinner';
import TextInput from '../TextInput/TextInput';
import PropTypes from 'prop-types';
import styles from './TopicDetailsForm.module.css';
import { useFormContext } from 'react-hook-form';
import hasAlphaNumericCharacter from '@/validators/alpha-numeric';
import isProfane from '@/validators/profanity';

const TopicDetailsForm = ({
  onSubmit,
  onInvalid,
  cancelHref,
  errorMessage,
  nameFieldName,
}) => {
  const {
    register,
    handleSubmit,
    formState: { isSubmitting, errors },
  } = useFormContext();

  const nameFieldOptions = {
    required: 'Required',
    maxLength: {
      value: 50,
      message: 'Must be less than 50 characters',
    },
    validate: {
      hasAlphaNumericCharacter: (name) =>
        hasAlphaNumericCharacter(name) ||
        'Must contain an alpha-numeric character',
      isNotProfane: (name) => !isProfane(name) || 'Must not contain profanity',
    },
  };
  const nameFieldErrorMessage = errors[nameFieldName]?.message;

  return (
    <div className={styles.TopicDetailsForm} data-testid="TopicDetailsForm">
      <form onSubmit={handleSubmit(onSubmit, onInvalid)}>
        {errorMessage && (
          <div className="mb-10">
            <ErrorAlert border={true} scrollTo={!!errorMessage}>
              {errorMessage}
            </ErrorAlert>
          </div>
        )}

        <div>
          <TextInput
            name={nameFieldName}
            label="Name"
            errorMessage={nameFieldErrorMessage}
            autoComplete="off"
            {...register(nameFieldName, nameFieldOptions)}
          />
        </div>

        <div className="mt-10 flex flex-col sm:flex-row">
          <button
            className="btn btn-primary w-100"
            type="submit"
            disabled={isSubmitting}
          >
            Submit
          </button>

          <Link href={cancelHref}>
            <a className="mt-3 text-center btn btn-cancel-outline w-100 sm:mt-0 sm:ml-3">
              Cancel
            </a>
          </Link>

          {isSubmitting && (
            <div className="mt-5 sm:mt-0 sm:ml-3 self-center">
              <LoadingSpinner size={30} />
            </div>
          )}
        </div>
      </form>
    </div>
  );
};

TopicDetailsForm.propTypes = {
  onSubmit: PropTypes.func,
  onInvalid: PropTypes.func,
  cancelHref: PropTypes.string,
  errorMessage: PropTypes.string,
  nameFieldName: PropTypes.string,
};

TopicDetailsForm.defaultProps = {};

export default TopicDetailsForm;
