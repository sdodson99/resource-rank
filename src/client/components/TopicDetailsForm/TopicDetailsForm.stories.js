import React from 'react';
import { FormProvider, useForm } from 'react-hook-form';
import TopicDetailsForm from './TopicDetailsForm';

export default {
  component: TopicDetailsForm,
  title: 'Components/TopicDetailsForm',
  // Fixes a storybook issue for maximum call stack exceeding.
  parameters: { docs: { source: { type: 'code' } } },
  decorators: [
    (Story) => {
      const methods = useForm({
        mode: 'onBlur',
        reValidateMode: 'onChange',
      });

      return (
        <FormProvider {...methods}>
          <Story />
        </FormProvider>
      );
    },
  ],
};

const Template = (args) => <TopicDetailsForm {...args} />;

export const Default = Template.bind({});
Default.args = {
  nameFieldName: 'name',
  cancelHref: '/',
};

export const Submitting = Template.bind({});
Submitting.args = {
  ...Default.args,
  isSubmitting: true,
};

export const Error = Template.bind({});
Error.args = {
  ...Default.args,
  errorMessage: 'Failed to submit form.',
};
