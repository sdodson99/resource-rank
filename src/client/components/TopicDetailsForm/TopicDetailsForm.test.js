import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { createRenderer } from 'react-test-renderer/shallow';
import TopicDetailsForm from './TopicDetailsForm';
import { useFormContext } from 'react-hook-form';
import hasAlphaNumericCharacter from '@/validators/alpha-numeric';
import isProfane from '@/validators/profanity';

jest.mock('react-hook-form');
jest.mock('@/validators/alpha-numeric');
jest.mock('@/validators/profanity');

describe('<TopicDetailsForm />', () => {
  let props;
  let formContext;

  beforeEach(() => {
    props = {
      onSubmit: jest.fn(),
      onInvalid: jest.fn(),
      cancelHref: '/back',
      errorMessage: '',
      nameFieldName: 'NAME',
    };

    formContext = {
      handleSubmit: jest.fn(),
      register: jest.fn(),
      formState: {
        errors: {},
      },
    };
    useFormContext.mockReturnValue(formContext);
  });

  afterEach(() => {
    useFormContext.mockReset();
    hasAlphaNumericCharacter.mockReset();
    isProfane.mockReset();
  });

  it('should mount', () => {
    render(<TopicDetailsForm {...props} />);

    const topicDetailsForm = screen.getByTestId('TopicDetailsForm');

    expect(topicDetailsForm).toBeInTheDocument();
  });

  describe('name field', () => {
    let actualNameFieldName;
    let actualNameFieldOptions;

    beforeEach(() => {
      render(<TopicDetailsForm {...props} />);

      actualNameFieldName = formContext.register.mock.calls[0][0];
      actualNameFieldOptions = formContext.register.mock.calls[0][1];
    });

    it('should register', () => {
      expect(actualNameFieldName).toBe(props.nameFieldName);
    });

    it('should register with required validation', () => {
      expect(actualNameFieldOptions.required).toBeDefined();
    });

    it('should register with max length validation', () => {
      expect(actualNameFieldOptions.maxLength.value).toBe(50);
    });

    it('should register with alpha-numeric validation', () => {
      hasAlphaNumericCharacter.mockReturnValue(false);

      const message =
        actualNameFieldOptions.validate.hasAlphaNumericCharacter('value');

      expect(message).toBeDefined();
    });

    it('should register with profanity validation', () => {
      isProfane.mockReturnValue(true);

      const message = actualNameFieldOptions.validate.isNotProfane('value');

      expect(message).toBeDefined();
    });
  });

  it('should render correctly', () => {
    const component = createRenderer().render(<TopicDetailsForm {...props} />);

    expect(component).toMatchSnapshot();
  });

  it('should render correctly with error message', () => {
    props.errorMessage = 'Error message.';

    const component = createRenderer().render(<TopicDetailsForm {...props} />);

    expect(component).toMatchSnapshot();
  });

  it('should render correctly when submitting', () => {
    props.isSubmitting = true;

    const component = createRenderer().render(<TopicDetailsForm {...props} />);

    expect(component).toMatchSnapshot();
  });
});
