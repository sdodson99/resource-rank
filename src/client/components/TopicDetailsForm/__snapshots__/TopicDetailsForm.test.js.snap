// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TopicDetailsForm /> should render correctly 1`] = `
<div
  className="TopicDetailsForm"
  data-testid="TopicDetailsForm"
>
  <form>
    
    <div>
      <ForwardRef(WrappedTextInput)
        autoComplete="off"
        label="Name"
        name="NAME"
      />
    </div>
    <div
      className="mt-10 flex flex-col sm:flex-row"
    >
      <button
        className="btn btn-primary w-100"
        disabled={false}
        type="submit"
      >
        Submit
      </button>
      <Link
        href="/back"
      >
        <a
          className="mt-3 text-center btn btn-cancel-outline w-100 sm:mt-0 sm:ml-3"
        >
          Cancel
        </a>
      </Link>
    </div>
  </form>
</div>
`;

exports[`<TopicDetailsForm /> should render correctly when submitting 1`] = `
<div
  className="TopicDetailsForm"
  data-testid="TopicDetailsForm"
>
  <form>
    
    <div>
      <ForwardRef(WrappedTextInput)
        autoComplete="off"
        label="Name"
        name="NAME"
      />
    </div>
    <div
      className="mt-10 flex flex-col sm:flex-row"
    >
      <button
        className="btn btn-primary w-100"
        disabled={true}
        type="submit"
      >
        Submit
      </button>
      <Link
        href="/back"
      >
        <a
          className="mt-3 text-center btn btn-cancel-outline w-100 sm:mt-0 sm:ml-3"
        >
          Cancel
        </a>
      </Link>
      <div
        className="mt-5 sm:mt-0 sm:ml-3 self-center"
      >
        <LoadingSpinner
          color="#292929"
          size={30}
        />
      </div>
    </div>
  </form>
</div>
`;

exports[`<TopicDetailsForm /> should render correctly with error message 1`] = `
<div
  className="TopicDetailsForm"
  data-testid="TopicDetailsForm"
>
  <form>
    <div
      className="mb-10"
    >
      <ErrorAlert
        border={true}
        scrollTo={true}
      >
        Error message.
      </ErrorAlert>
    </div>
    <div>
      <ForwardRef(WrappedTextInput)
        autoComplete="off"
        label="Name"
        name="NAME"
      />
    </div>
    <div
      className="mt-10 flex flex-col sm:flex-row"
    >
      <button
        className="btn btn-primary w-100"
        disabled={false}
        type="submit"
      >
        Submit
      </button>
      <Link
        href="/back"
      >
        <a
          className="mt-3 text-center btn btn-cancel-outline w-100 sm:mt-0 sm:ml-3"
        >
          Cancel
        </a>
      </Link>
    </div>
  </form>
</div>
`;
